### ISSUE TITLE
Conditional dom constraint produces symmetries

### ISSUE THREAD

**Issue by MaxOstrowski** [potassco/clingcon#43]
```
{var("z")}.
:-not var("z"), &sum{"x"} = 5.
:-not var("z"), &sum{"x"} = 6.
:-not var("z"), &sum{"x"} = 10.
&dom{5; 6; 10}=("x"):-var("z").
&dom{(5..10)}=("x").
```
produces 15 results instead of 6 (all correct).
Some aux variables are not constrained enough.

**Comment by MaxOstrowski** [comment]
I fear that I do not completely understand the https://github.com/potassco/clingcon/blob/wip/libclingcon/src/solver.cc#L1068 function, but it does not seem that there are any aux literals introduced.
It should all be order literals, right ?

**Comment by rkaminsk** [comment]
The &dom constraints work fine. It is something else.

**Comment by MaxOstrowski** [comment]
Also holds for:
```
{var("z")}.
:-not var("z"), &sum{"x"} = 5.
:-not var("z"), &sum{"x"} = 10.
&sum{5}=("x"):-var("z").
&dom{(5..10)}=("x").
```

**Comment by rkaminsk** [comment]
Something goes wrong with the `!=` in the head.
```{var("z")}.
&sum{"x"} != 5 :- not var("z").
%:-not var("z"), &sum{"x"} = 5.
&dom{5; 6} = "x" :- var("z").
&dom{5..7} = "x".
```

**Comment by rkaminsk** [comment]
It really looks like it is both the `!=` and `=` translation in the head that is buggy. And now the bad news. This is the part you implemented. :wink: Maybe you did not consider the non-strict case correctly?
```
{var("z")}.
%{ does not work
%&sum{"x"} != 5 :- not var("z").
%}
%{ works
a | b :- not var("z").
&sum{"x"} < 5 :- a.
&sum{"x"} > 5 :- b.
%}
%{ works with --no-shift-constraints
%:-not var("z"), &sum{"x"} = 5.
%}
&dom{5; 6} = "x" :- var("z").
&dom{5..7} = "x".
```

**Comment by rkaminsk** [comment]
```cpp
        if (!builder.add_clause({a, b, -literal})) {
            return false;
        }
        if (!builder.add_clause({-a, -b})) {
            return false;
        }

        if (!normalize_constraint(builder, a, *elems, "<", rhs, false)) {
            return false;
        }
        if (!normalize_constraint(builder, b, *elems, ">", rhs, false)) {
            return false;
        }
```

With the above code, if the literal is false, the following cases are permitted:
- `a` is false and `b` is false
- `a` is false and `b` is true
- `a` is true and `b` is false

If `var(z)` is true, then `literal` is false. If `x=6`, then `b` is false and `b` is true both won't cause a conflict and there are too many solutions. This might also affect clingo-dl. One fix could be to set both `a` and `b` to false if the literal is false. But maybe we can optimize in the strict case by not adding these clauses. I did not fully think the thing through to be honest.

**Comment by MaxOstrowski** [comment]
I'm already on it, just had lunchbreak :grin:

**Comment by MaxOstrowski** [comment]
fixed in wip dd0df67

**Comment by MaxOstrowski** [comment]
Timings improved with this patch. Not drastically, but noticeable.
